src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+using System.Linq;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class NullCheckRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        private readonly MethodSymbol _method;
+        private readonly SyntheticBoundNodeFactory _fact;
+        private NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundStatement Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            if (!method.Parameters.Any(x => x is SourceParameterSymbolBase param
+                                            && param.IsNullChecked))
+            {
+                return body;
+            }
+
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return (BoundStatement)rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            return AddNullChecksToBody(node);
+        }
+
+        private BoundNode AddNullChecksToBody(BoundBlock body)
+        {
+            var statementList = ArrayBuilder<BoundStatement>.GetInstance();
+            foreach (ParameterSymbol x in _method.Parameters)
+            {
+                if (x is SourceParameterSymbolBase param && param.IsNullChecked)
+                {
+                    var constructedIf = ConstructIfStatementForParameter(body, param);
```

Comment:
>ConstructIfStatementForParameter [](start = 40, length = 32)

We should not generate this code if `parameter.Type` is a value type and not `Nullable<>`.

Reply:
It seems like we shouldn't make it to lowering when the "null-checked parameter" is a non-nullable value type.

Reply:
The spec says this:

> The check will be specifically for reference equality to null, it does not invoke == or any user defined operators. This also means the ! operator can only be added to parameters whose type can be tested for equality against null. This means it can't be used on a parameter whose type is known to be a value type.

However it seems like you could get into this scenario via generic substitution. So it should probably be handled and exercised that way.

Reply:
We should test cases of substitution:
```C#
abstract class A<T>
{
    internal abstract void M<U>(U u) where U : T;
}
class B1<T> : A<T> where T : class
{
    internal override void M<U>(U u!) { }
}
class B2<T> : A<T> where T : struct
{
    internal override void M<U>(U u!) { }
}
class B3 : A<int>
{
    internal override void M<U>(U u!) { }
}
class B4 : A<int?>
{
    internal override void M<U>(U u!) { }
}
class B5 : A<object>
{
    internal override void M<U>(U u!) { }
}
```

---
In reply to: [295530405](https://github.com/dotnet/roslyn/pull/36471#discussion_r295530405) [](ancestors = 295530405)

Reply:
Added assert, will work on tests tomorrow morning.

Reply:
Honestly looking at the spec again it doesn't appear to me that `int? myParam!` is meant to be allowed. However I think we should proceed for now as if it is allowed, because changing course also requires specifying the behavior of `void M<T>(T t!) { }` with an invocation `M<int?>(null);`.

---
In reply to: [295537131](https://github.com/dotnet/roslyn/pull/36471#discussion_r295537131) [](ancestors = 295537131,295530405)